### ISSUE TITLE
Date Picker Component

### ISSUE THREAD

**Issue by gselltho** [dynatrace-oss/barista#894]
# Feature Request

## Summary

Add a dedicated component for picking dates. Similar to the date picker component of ng material: https://material.angular.io/components/datepicker/overview

## Feature Description
The component should support these attributes:
- date: Date; Binding for the value of the component
- format: String; Date format, something like: "YYYY/MM/DD" 
- disabled: boolean; is the component enabled or disabled
- changed: event; fires when value changes
- rangeFrom: Date; selectable range of date
- rangeTo: Date; selectable range of date

to start with, it will be sufficient to provide a basic version: input field with an button next do it to open the date selector overlay (as to see in the attached screenshot)

## Attachments
Screenshot of Material Date Picker:
![datepicker](https://user-images.githubusercontent.com/28366423/79730525-42f05f80-82f1-11ea-957b-32571aeeafc4.png)

**Comment by lukasholzer** [comment]
@gselltho Thank you very much for opening this feature request!
I think on this topic we need a close collaboration with our @dynatrace-oss/ux-core team. Maybe they have already created some mocks for this topic or they can provide us with some insights.

If they have already some mocks or designs, this issue would be a good starting point to collect all necessary information and API proposals.

**Comment by kathrin-aigner** [comment]
I'm not sure for which purpose the date picker is needed but Arkadiusz Lipiarz is working together with his team for a long time now on a date picker we want to use as Global timeframe selector 
![image](https://user-images.githubusercontent.com/57905306/79749293-3c70e080-830f-11ea-8b3a-96fcebd96309.png) it's already implemented in our demo environment. Ithink in this case it would be best to talk to him.

**Comment by ffriedl89** [comment]
We had the discussion in the past and @ursula-wieshofer did some evaluation in the product, and we have one or two datepickers in total. This meant that this was not high on the priority list. 

@gselltho I think it would make sense to discuss this further and target an addition to the library. 
I think we can take a lot of learnings from @angular/components and do a simplified version. 
Would you be able to create a more detailed API Proposal for it?

**Comment by arkadiuszlipiarz** [comment]
Hello there!

So it seems that the feature my team created would fit use cases @gselltho listed and much much more. Besides the top of our site, it is also used by Dashboard team as a local component. We prepared it both in Angular and GWT. Reach out to me on Slack if you are interested.

But... it seems that Ryan and his team were preparing a super similar solution to what you shared. Here is how it looks:

![Screenshot 2020-04-21 at 10 56 23](https://user-images.githubusercontent.com/26602747/79846731-1440ba80-83bf-11ea-83a9-0c2abf4da904.png)

Code owner is @areknow and he can provide you with more details.

**Comment by gselltho** [comment]
Hi guys, i'm not sure whats next. The component of Ryans team is basically the angular material one and exact what i need. But I would say, since it is a very basic component, its home should be barista components and as it seems, more teams could use it.

@ffriedl89: of course, i can create a more detailed API description. At least for a limited feature set for a first implementation. I will post it here in the next couple of days

**Comment by areknow** [comment]
Hi @gselltho, I am the developer of the date picker shared component that resides in the spine-x repo. I have already spoken to the barista team about transferring it to the barista components lib, but unfortunately it has a hard dependency on several material packages which is not something that the team would like to add to the project. 

Angular CDK does not include a date picker which means this component would have to be written from the ground up. 

Is there an issue with you using the date picker shared component from the mono repo?

Cheers

**Comment by gselltho** [comment]
Hi @areknow, your component is great! I just miss Angulars Reactive Form feature (ControlValueAccessor) and, so it seems, it has to be a DtFormFieldControl to use it seamless within Barista and Angular Reactive forms.

**Comment by ffriedl89** [comment]
@gselltho Thanks for reaching out again. It's exactly as @areknow outlines. His component is a restyled material component which introduces a dependency on material and other util libraries from the angular/components repo. It also comes with a hard dependency to moment.js which is huge and not tree shakable. 
Unfortunately, the CDK does not come with a datepicker yet that is not styled the material way. 
Up until now the datepicker was not a component many teams needed since we don't have any datepickers except the global timeframe selector in the product. 

If you need the component it would be awesome if we can work together on contributing the component to the library. This process starts by creating the API proposal for it here in this issue :) 
Once this is done we can help you get started and guide you through the development process and we will help you along the way. 

Right now the barista team needs help from contributors to add this component to the library.
Any help here is highly appreciated

**Comment by thomaspink** [comment]
Hey, will take care of it as the setting VCT also needs a date/timepicker now. In the following comments i will try to collect all requirements and the propose an API. 

If you also have and features/requirements that should be in there, that are not yet listed, please let me know.

**Comment by thomaspink** [comment]
# Requirements:

## General Requirements
- **Selecting a date**: The user must be able to select a date
- **Selecting a time**: In addition to selecting a date the user must also be able to select an additional time.
- **Prefill
- **Disabled**: The consumer may disable the whole datepicker.
- **Selectable range**: It should be possible to define the range of dates the user can select (can be implemented via min/max inputs)

## Technical Requirements
- **NgModel compatible**: The datepicker must be compatible with angular forms.
- **Form-field compatible**: The datepicker must be a valid form-field-control.
- **Modularity**: The datepicker should consist of several sub-components/directives/utilities so the consumers can also use just parts instead of the full datepicker component (e.g. The month-view for timeframe selector)
- **Native Date object**: Dates should be represented by the native `Date` object/class (including the whole consumer facing API). Not by some abstract custom implementation.  Additional utility functions/services (`DateAdapter`) might be needed.

Optional Requirements *(Not sure if needed)*:
- **Limit selectable dates**: It should be possible to further limit (extending the min/max requirements) the dates that are selectable (can be implemented by a function the consumer can implement)
- **Ranges**: In addition to select a single date it should also be possible to select a range of dates (e.g. 1.1.2020 - 10.1.2020)

**Comment by thomaspink** [comment]
Please post below the requirements/feature the datepicker should provide. I will add them in the list of the previous comment.

**Comment by gselltho** [comment]
@thomaspink Thank you for taking care of this 👍 
Your requirement description has already everything i would need. I have just a question regarding:
`Selecting a time: In addition to selecting a date the user must also be able to select an additional time.`
The time is meant as optional, i guess? Will I be able to chose what is selectable: only date, only time, or both in combination?

**Comment by thomaspink** [comment]
@gselltho Thanks for the feedback 👍 

The idea was to either have date and time as default, but you can disable the time part to make it date only, or the other way round. Not sure what the best way is (do more consumers need only a date selectors or the date/time one).

**Comment by areknow** [comment]
It would be nice to be have control over timezone. Some of our use cases in spine X require UTC for interacting with certain legacy APIs. 

Also like @gselltho said, having the option to have just date alone would be nice. 

Thanks for this project!

**Comment by toddbaert** [comment]
Will this include predefined relative dates? As in "now", "two weeks ago", etc?

**Comment by pieniazek** [comment]
One more requirement (~ GTS):
- We'd need light and dark version

Nice to have:
- on mobile browsers -> fallback to native date/time picking controls

**Comment by thomaspink** [comment]
@toddbaert Apart from the global timeframe selector is this also a usecase for this general datepicker?

@pieniazek added to list

**Comment by ursula-wieshofer** [comment]
@thomaspink I think because of consistency reasons and to make the handling of the date picker easier for the user we should stick to the part of the global timeframe selector that also includes the calendar. The basic functionality and handling should be the same as well as the style. 
This will support the user when dealing with it as he/she is used to the functionality already.
![Screenshot 2020-06-25 at 08 22 56](https://user-images.githubusercontent.com/36620967/85716598-e8a6bd00-b6ec-11ea-9abc-859198a4842c.png)

**Comment by toddbaert** [comment]
@thomaspink  it wont be a deal-breaker for user, but we are exposing some UIs soon for which the predefined-relative dates could be useful.

**Comment by thomaspink** [comment]
@toddbaert do you have a list of the supported values? I will add it to the nice to have list if its ok for you

**Comment by thomaspink** [comment]
For the datepicker we would also need some kind of a service (similar to angular materials data-adapter) that provides us with the functionality and information so we can display the dates and the correct namings (including weekday names) on the one hand and calculate/work with native dates on the other hand. This "service" should then also be available to our consumers so they can work with dates the same way and we can share this functionality.

For that we have at least 2 options (if you have more ideas i would appreciate it if you would post them below):

1. Create a NativeDateAdapter which combines native date APIs (`Intl`, `Date`) and utility functions (such as `addCalendarDays`). This Adapter would be very similar to Angular Materials [NativeDateAdapter](https://github.com/angular/components/blob/master/src/material/core/datetime/native-date-adapter.ts). But for our usecases we do not need the fallbacks when `Intl` is not available, as all our supported browser have already `Intl' built in. We would also not need the possibility to have DateAdapter for different libraries as we only work with the native `Date` API.
2. Use an external library such as [`date-fns`](https://date-fns.org/) and wrap it into a service (DateAdapter) we then provide to our customers.

Right now i would tend more toward option 1 as it uses heavily the native APIs without the need for an additional library but please also post your ideas, findings and comments :)

**Comment by tomheller** [comment]
@thomaspink I agree that we should create such an adapter service. I know that I stated previously, that I would love to use the `Intl` and `Date` natives for this one, but I double checked the [support](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat) for `Intl.DateFormat` and I'm not super happy about it. Especially the partially missing Range formats are a bit sad. 

Seeing that `date-fns` is [treeshakable and does not look to be a large library in general](https://bundlephobia.com/result?p=date-fns@2.14.0), I would now opt to use `date-fns`. 
Additionally, date and time math can be really hard, even with a lot of native support when dealing with client timezones, etc...

As I see that you already have created a draft PR for the date-picker. Is there already an API that you can propose, regarding the component?

**Comment by thomaspink** [comment]
This PR is wip and i did not yet tackle the tasks that requires such a service. Still not sure which functionality we are gonna need for this service but i will start with a basic one and expand it on the way.

**Comment by piotrl** [comment]
### Time zones
Recently we dropped this requirement. There are many tiny yet complex details to consider with it that are hard to explain to user for UX perspective.

### Date-manipulation library
date-fns is fantastic API for time-shifting operations like `addMonth`, `getDaysInMonth`, `isToday`, `startOfTheWeek` etc. Tree-shake'ability makes it lightweight usage. Probably it'd cost as much weight as the adapter abstraction :P

On the other hand, library-consumer would have mixed feelings if they need to include `date-fns` when they have moment or luxon -> Here adapter would be useful.

Note about time zones: `date-fns` won't be suitable if you'd like to implement full time zones support (even with `date-fns-tz` which solves only formatting/parsing issues, but no date-manipulation scenarios). Recently we considered to use `luxon` for this purpose, at least until [Temporal specification](https://github.com/tc39/proposal-temporal) is not finished and adopted by browsers.

### Formatting 
* Do you know why Angular Material datepicker does not use simply DatePicker?
* Missing range formatting from `Intl.DateFormat`: Perhaps we could implement our own fallback or use polyfil? In product we already have something in place.
* Note if we won't use DatePicker for formatting, it would be good to support [LOCALE_ID](https://angular.io/api/core/LOCALE_ID) token to control how text is translated.

**Comment by thomaspink** [comment]
Thanks @piotrl 

As far as i understand it, the DatePipe would be only useful for displaying the finally selected date, but not when calculating/displaying things like: All days of a certain month.

**Comment by thomaspink** [comment]
Another positive side effect would be that if we stick closely to the api of the material date-adapter, and they are gonna move it to the cdk, we could easily adapt and change it.

**Comment by thomaspink** [comment]
@areknow 
> It would be nice to be have control over timezone. Some of our use cases in spine X require UTC for interacting with certain legacy APIs.

Having a date-adapter would also solve this requirement, as the datepicker treats dates as whatever generic object the adapter provides. This can be a native Date, a Moment wrapper or date that always has a specific timezone.

**Comment by piotrl** [comment]
> If we gonna implement a date-adapter would the cluster also use it as the central date entry point?

I'm open to consider it. My initial thoughts are like raw Date object is used in 600+ places already (explicitly, can't count implicit usages), most of them won't be anyhow connected to calendar or even close to calendar as most of dates comes from backend. Using adapter for everything for sake of consistency with calendar might be overwhelming.

So if we decide to go with adapter, there are still 2 ways to go:
1. Provide adapter in module configuration ONLY for internal calendar date-operations, model (input/output) would still base on native dates
2. Provide adapter for both internal usages + input/output API.

2nd option is more *invasive* in terms of use-cases to think of, requires well thought API of adapter.
1st option could be easier to start with and could evolve to 2nd if needed.

**Comment by lukasholzer** [comment]
Moved to APM-266081
